pkg/kubecost/summaryallocation.go

```diff
@@ -0,0 +1,1155 @@
+			return nil, err
+		}
+	}
+
+	return acc, nil
+}
+
+// AggregateBy aggregates the Allocations in the given AllocationSet by the given
+// AllocationProperty. This will only be legal if the AllocationSet is divisible by the
+// given AllocationProperty; e.g. Containers can be divided by Namespace, but not vice-a-versa.
+func (sas *SummaryAllocationSet) AggregateBy(aggregateBy []string, options *AllocationAggregationOptions) error {
+	// The order of operations for aggregating allocations is as follows:
+	//
+	//  #. Partition external, idle, and shared allocations into separate sets.
+	//     Also, create the aggSet into which the results will be aggregated.
+	//
+	//  #. Retrieve pre-computed allocation resource totals, which are to be
+	//     used to compute idle coefficients, based on the ratio of any given
+	//     allocation's per-resource cost to the allocated per-resource totals.
+	//
+	//  #. Distribute idle allocations according to the idle coefficients
+	//
+	//  #. Generate aggregation key and insert allocation into the output set
+	//
+	//  #. If idle is shared and other allocations are shared, it's probable
+	//     that some amount of idle cost will be shared with a shared resource.
+	//     Distribute that idle allocation, if it exists, to the respective
+	//     shared allocations before sharing them with the aggregated
+	//     allocations.
+	//
+	//  #. Apply idle filtration
+	//
+	//  #. Distribute shared allocations
+	//
+	//  #. Distribute shared tenancy costs
+	//
+	//  TODO
+	//  #. If there are external allocations that can be aggregated into
+	//     the output (i.e. they can be used to generate a valid key for
+	//     the given properties) then aggregate; otherwise... ignore them?
+	//
+	//  #. If the merge idle option is enabled, merge any remaining idle
+	//     allocations into a single idle allocation. If there was any idle
+	//	   whose costs were not distributed because there was no usage of a
+	//     specific resource type, re-add the idle to the aggregation with
+	//     only that type.
+	//
+	//  TODO
+	//  #. Distribute any undistributed idle, in the case that idle
+	//     coefficients end up being zero and some idle is not shared.
+
+	if sas == nil || len(sas.SummaryAllocations) == 0 {
+		return nil
+	}
+
+	if sas.Window.IsOpen() {
+		return errors.New("cannot aggregate a SummaryAllocationSet with an open window")
+	}
+
+	if options == nil {
+		options = &AllocationAggregationOptions{}
+	}
+
+	if options.LabelConfig == nil {
+		options.LabelConfig = NewLabelConfig()
+	}
+
+	// Check if we have any work to do; if not, then early return. If
+	// aggregateBy is nil, we don't aggregate anything. On the other hand,
+	// an empty slice implies that we should aggregate everything. (See
+	// generateKey for why that makes sense.)
+	shouldAggregate := aggregateBy != nil
+	shouldShare := len(options.SharedHourlyCosts) > 0 || len(options.ShareFuncs) > 0
+	if !shouldAggregate && !shouldShare {
+		return nil
+	}
```

Comment:
Does this have identical behavior to `AllocationSet` in which an empty aggregation is not the same as the "default" (from the HTTP API's perspective) aggregation of `cluster/node/namespace/...`?

Reply:
The behavior should match, but I'm not sure that premise is the case. Will double check, but I think they're the same in both `Allocation` and `SummaryAllocation`. Possibly mistaken, of course.

Reply:
I think I misunderstood your comment. Yes, the two APIs share behavior. In fact, the even share that precise code path now in `package main`'s `ParseAllocationRequest`